apollo-federation/src/query_graph/extract_subgraphs_from_supergraph.rs

```diff
@@ -2094,10 +2104,181 @@ fn maybe_dump_subgraph_schema(subgraph: FederationSubgraph, message: &mut String
+                let subgraph = get_subgraph(
+                    subgraphs,
+                    graph_enum_value_name_to_subgraph_name,
+                    &subgraph_enum_value,
+                )?;
+
+                schema_def_pos.insert_directive(
+                    &mut subgraph.schema,
+                    Component::new(link_directive.clone()),
+                )?;
+
+                // TODO: add imported definitions from relevant specs
+            }
+        }
+
+        // Other directives are added normally.
+        for (directive, subgraph_enum_values) in others {
+            for subgraph_enum_value in subgraph_enum_values {
+                let subgraph = get_subgraph(
+                    subgraphs,
+                    graph_enum_value_name_to_subgraph_name,
+                    &subgraph_enum_value,
+                )?;
+
+                schema_def_pos
+                    .insert_directive(&mut subgraph.schema, Component::new(directive.clone()))?;
+            }
+        }
+    }
+
+    for object_field_pos in &join_directives.object_fields {
+        let object_field = object_field_pos.get(supergraph_schema.schema())?;
+        let directives = object_field
+            .directives
+            .iter()
+            .filter_map(|d| {
+                if d.name == JOIN_DIRECTIVE {
+                    Some(join_directive_to_real_directive(d))
+                } else {
+                    None
+                }
+            })
+            .collect_vec();
+
+        for (directive, subgraph_enum_values) in directives {
+            for subgraph_enum_value in subgraph_enum_values {
+                let subgraph = get_subgraph(
+                    subgraphs,
+                    graph_enum_value_name_to_subgraph_name,
+                    &subgraph_enum_value,
+                )?;
+
+                object_field_pos
+                    .insert_directive(&mut subgraph.schema, Node::new(directive.clone()))?;
+            }
+        }
+    }
+
+    // TODO
+    // - join_directives.directive_arguments
+    // - join_directives.enum_types
+    // - join_directives.enum_values
+    // - join_directives.input_object_fields
+    // - join_directives.input_object_types
+    // - join_directives.interface_field_arguments
+    // - join_directives.interface_fields
+    // - join_directives.interface_types
+    // - join_directives.object_field_arguments
+    // - join_directives.object_types
+    // - join_directives.scalar_types
+    // - join_directives.union_types
+
+    Ok(())
+}
+
+fn join_directive_to_real_directive(directive: &Node<Directive>) -> (Directive, Vec<Name>) {
+    let subgraph_enum_values = directive
+        .argument_by_name("graphs")
+        .and_then(|arg| arg.as_list())
+        .map(|list| {
+            list.iter()
+                .map(|node| {
+                    Name::new(
+                        node.as_enum()
+                            .expect("join__directive(graphs:) value is not an enum")
```

Comment:
the message of `.expect()` should be explaining why the Result/Option of the value should have been there ([guidance from rust docs](https://doc.rust-lang.org/std/result/enum.Result.html#recommended-message-style)). All of these are worded in an `.assert()` style when it's not actually an assert. These should read something like "join__directive(graphs:) value is an enum" or something of the sort. There are a bunch of good examples in `apollo-router/src/json_ext.rs` with expect statements.